File: Leetcode/sbgo/Leetcode397-整数替换.go
package sbgo

func integerReplacement(n int) int {
	f := map[int]int{}
	var dp func(int) int
	dp = func(n int) int {
		if _, ok := f[n]; ok {
			return f[n]
		}
		if n == 1 {
			return 0
		}
		if n%2 == 0 {
			f[n] = dp(n/2) + 1
			return f[n]
		}
		f[n] = min(dp(n+1), dp(n-1)) + 1
		return f[n]
	}
	return dp(n)
}

File: Leetcode/JS/Leetcode397-整数替换.js
var integerReplacement = function(n) {
    let f = new Map();
    let dp = function(n) {
        if(f.has(n)) return f[n];
        if(n == 1) return 0;
        if(n % 2 == 0){
            return f[n] = dp(n / 2) + 1;   
        }
        return f[n] = Math.min(dp(n + 1),dp(n - 1)) + 1;
    }
    return dp(n);
};